### ISSUE TITLE
Subset assignment splits frames of the same video across train/val/test, inflating evaluation metrics

### ISSUE THREAD

**Issue by shashb27** [open-edge-platform/geti#7619]
**Describe the bug**

First of all — thank you for open-sourcing Geti; the API-first workflow has been a joy to drive end to end. While validating models on held-out footage we ran into something we believe is worth your attention.

`SubsetAssigner` (`application/backend/app/services/subset_assignment/assigner.py`) stratifies dataset items by label only — `DatasetItemWithLabels` carries no media identity. When a project's annotations are video frames, frames of the *same video* are distributed freely across TRAINING, VALIDATION and TESTING.

Since temporally adjacent frames of one video are near-duplicates, the test subset effectively contains copies of the training data, and evaluation metrics come out inflated — with no indication to the user. We verified the assigner is unchanged at `develop` @ `e75ff15f1`.

**Steps to Reproduce**

1. Create a detection project and upload a single video; annotate ~30 frames spread across it (one label).
2. Train any detection architecture (we tried YOLOX-Tiny and RF-DETR-M, on two different datasets).
3. Observe test metrics: we consistently got **mAP@0.5 = 1.0**.
4. Run the trained model on held-out footage of the same object (different scene/lighting): we measured as low as **48% frame-level recall** — while the in-app metric reads 1.0 with no caveat.

**Environment:**

- Geti app / library version: 3.1.0 (also verified code path unchanged on `develop` @ e75ff15f1)
- Installation mode: Docker (pre-built `ghcr.io/open-edge-platform/geti-cuda`)
- Hardware spec: RTX 5090 (32 GB), 64-thread CPU
- OS: Ubuntu 22.04 host
- GPU drivers: CUDA driver 580.178.04

**Suggested direction (happy to contribute)**

The schema already links frames to their parent video (`media.video_id`), so we prototyped:
1. Group-aware assignment — frames of the same video stay within one subset (group-level iterative stratification) whenever ≥3 distinct media groups exist;
2. A logged warning when leakage is unavoidable (e.g. all annotations come from a single video — arguably the most common beginner flow), noting that metrics likely overestimate real-world performance.

We have this working with unit + integration tests passing and would be glad to open a PR — equally happy to adapt to whatever direction the maintainers prefer. Thanks for taking a look!

**Comment by leoll2** [comment]
Thanks for your feedback @shashb27 , you are right that the evaluation metrics can become overly optimistic when the validation and testing subsets contain video frames that are nearly identical to samples of the training set.

When there are 3+ annotated videos, your proposed approach (assign all frames of one video to the same subset) makes a lot of sense. The case with 1-2 videos is more tricky to handle in a general way, because the three subsets can't come from fully independent sets of frames. A simple heuristics to to ensure subsets that are as close as possible to independent samples could be splitting the video into contigous temporal fragments, e.g. :

```
00:00 ───────── 40:00 ─── 45:00 ── 50:00
        TRAIN         VAL      TEST
```
With this approach, however, frames around the boundaries (40:00, 45:00) would be still correlated. If there are enough annotated frames, one could just discard these near-boundary frames to avoid the problem altogether; if the number of annotated frames is small, though, it might not be viable.

Anyway, I will review your PR and leave additional feedback there, I think it's alerady a great starting point. I'll also bring this problem to the attention of the team if someone comes up with additional refinements. Thank you very much for your contribution!

**Comment by shashb27** [comment]
Thank you for the thoughtful response, @leoll2 — and for validating the problem so quickly!

The temporal-fragment idea for the 1–2 video case is elegant, especially with near-boundary frames discarded when the annotation budget allows. It would give single-video projects genuinely more independent subsets than the current frame-level split, and it composes naturally with the group-aware path (fragments become the groups).

Happy to take this in whatever direction works best for the team: keep the current PR scoped to group-aware assignment + the warning (with temporal fragments as a follow-up), or extend it now with the fragment fallback for the low-video-count case — whichever you prefer. Looking forward to your review feedback on #7621, and thanks again for engaging with this so constructively.
